lib/autokey/scripting/window.py

```diff
@@ -247,3 +247,73 @@ def _run_wmctrl(self, args):
             return 1, 'ERROR: Please install wmctrl'
 
         return returncode, output
+
+    def get_window_list(self, filter_desktop=None):
+        """
+        Returns a list of windows matching an optional desktop filter, requires C{wmctrl}!
+
+        Each list item consists of: C{[hexid, desktop, hostname, title]}
+
+        Where the C{hexid} is the ID used for some other functions (like L{import -window} from ImageMagick).
+
+        C{desktop} is the number of which desktop (sometimes called workspaces) the item appears upon.
+
+        C{hostname} is the hostname of your computer (I'm not sure why this is included as it seems to be the same for everything I have seen)
+        
+        C{title} is the title that you would usually see in your window manager of choice.
+
+        @param filter_desktop: String, (usually 0-n) to filter the windows by. Any window not on the given desktop will not be returned.
+        @return: C{[[hexid1, desktop1, hostname1, title1], [hexid2,desktop2,hostname2,title2], ...etc]}
+        Returns C{[]} if no windows are found.
+        """
+        windowList = self._run_wmctrl(["-l"])[1].split("\n")
+        output = []
+        for window in windowList:
+            info = window.split("  ")[1] #the remaining information
```

Comment:
Tidy up the inside of this loop with a function maybe? Something that just returns the list of useful info?

You can use `@staticmethod` above the function declaration to avoid having to include API documentation, I think.

If you're really keen, getting some tests written where you can would be great. Even if you mock out the external bits to return the expected results, that would still be super useful.

Reply:
I may have brought it up elsewhere but I think we should probably swap from using wmctrl to internal implementation because implementing reading output from the commandline is always going to be a pain to implement (and be easily broken if `wmctrl` changes anything about their output)

I say this understanding that the main thing we are lacking atm is developer time and re implementing existing features is not the best use of time

Reply:
yeah, reasonable point. One stop-gap to cover wmctrl changing output is to create some unit tests that assert its interface is what we expect it to be. Have you had a go with the testing system yet? It's very liberating once you get used to it.

Reply:
I've changed my mind about re-writing this loop (so ignore my comment about `@staticmethod`). Would still be nice to have some tests though.

I think the easiest way to start is to create `tests/scripting_api/test_window.py`. Copy the contents of `tests/scripting_api/test_macro.py` to get the relevant setup methods and imports, etc. (Feel free to change the copyright notice at the top to your name/username and email).

The [mock objects](https://docs.python.org/3/library/unittest.mock.html) and [patches](https://docs.python.org/3/library/unittest.mock.html#patch) that you can create are very useful. In this case, you might want a test body like 

```python
get_win_details = MagicMock(return_value=test_string)
with patch("self._run_wmctrl", get_win_details):
    output = engine.get_window_list()
    assert_equal(output, expected_output)
```

Where `test_string` is an example of what you might expect `self._run_wmctrl(["-l"])` to return in a real system, and `expected_output` is what you expect the function to return.

Create one of these for now, and when you've got the hang of it, I'll show you how to test multiple test strings in a single test. (Or play with the `@parameterize` decorator if you want to learn yourself).

Each test you write is a function in the file that starts with `test_`.